pkg/registry/core/service/storage/rest.go

```diff
@@ -541,33 +584,327 @@ func (r *REST) ConvertToTable(ctx context.Context, object runtime.Object, tableO
+				}
+				service.Spec.IPFamilies = append(service.Spec.IPFamilies, api.IPv6Protocol)
+			} else {
+				// first make sure that family(ip) is configured
+				if _, found := rs.serviceIPAllocatorsByFamily[api.IPv4Protocol]; !found {
+					el := field.ErrorList{field.Invalid(field.NewPath("spec", "clusterIPs").Index(i), service.Spec.ClusterIPs, "ClusterIP of IPv4 can not be used, IPv4 is not configured on cluster")}
+					return errors.NewInvalid(api.Kind("Service"), service.Name, el)
+				}
+				service.Spec.IPFamilies = append(service.Spec.IPFamilies, api.IPv4Protocol)
+			}
+		}
+	}
+
+	// default headless+selectorless
+	if len(service.Spec.ClusterIPs) > 0 && service.Spec.ClusterIPs[0] == api.ClusterIPNone && service.Spec.Selector == nil {
+		if service.Spec.IPFamilyPolicy == nil {
+			requireDualStack := api.RequireDualStack
+			service.Spec.IPFamilyPolicy = &requireDualStack
+		}
+
+		// if not set by user
+		if len(service.Spec.IPFamilies) == 0 {
+			service.Spec.IPFamilies = []api.IPFamily{rs.defaultServiceIPFamily}
+		}
+
+		// this follows headful services. With one exception on a single stack
+		// cluster the user is allowed to create headless services that has multi families
+		// the validation allows it
+		if len(service.Spec.IPFamilies) < 2 {
+			if *(service.Spec.IPFamilyPolicy) == api.RequireDualStack ||
+				(*(service.Spec.IPFamilyPolicy) == api.PreferDualStack && len(rs.serviceIPAllocatorsByFamily) == 2) {
+				// add the alt ipfamily
+				if service.Spec.IPFamilies[0] == api.IPv4Protocol {
+					service.Spec.IPFamilies = append(service.Spec.IPFamilies, api.IPv6Protocol)
+				} else {
+					service.Spec.IPFamilies = append(service.Spec.IPFamilies, api.IPv4Protocol)
+				}
+			}
+		}
+
+		// nothing more needed here
+		return nil
+	}
+
+	// ipfamily check
+	if len(service.Spec.ClusterIPs) > 0 && service.Spec.ClusterIPs[0] == api.ClusterIPNone && service.Spec.Selector == nil {
+		klog.V(4).Infof("service %v/%v is a headless service. no validation will be perfomed on its ipfamilies:%v", service.Namespace, service.Name, service.Spec.IPFamilies)
+	} else {
+		// the following applies on all type of services including headless w/ selector
+
+		// asking for dual stack on a none dual stack cluster
+		// should fail without assigning any family
+		if service.Spec.IPFamilyPolicy != nil && *(service.Spec.IPFamilyPolicy) == api.RequireDualStack && len(rs.serviceIPAllocatorsByFamily) < 2 {
+			el := field.ErrorList{field.Invalid(field.NewPath("spec", "ipFamilyPolicy"), service.Spec.IPFamilyPolicy, "Cluster is not configured for dual stack services")}
+			return errors.NewInvalid(api.Kind("Service"), service.Name, el)
+		}
+
+		//TODO(khenidak) this code, will become dead, because the above check will trigger before this one
+		// we need to make a choice which error we want to return (family or general "no dual stack for you"
+		// OR find a way to return multi validation error similar to validation package
```

Comment:
You can return multiple errors here; you just need to rearrange the code to allow adding more than one `field.Invalid()` to your `field.ErrorList` and then only call `errors.NewInvalid` at the end if `len(el) > 0`